libyul/backends/evm/StackHelpers.h

```diff
@@ -0,0 +1,408 @@
+		{
+			// We need another copy of this slot.
+			if (_ops.sourceMultiplicity(sourceOffset) > 0)
+			{
+				// If this slot occurs again later, we skip this occurrence.
+				if (ranges::any_of(
+					ranges::views::iota(sourceOffset + 1, _ops.sourceSize()),
+					[&](size_t _offset) { return _ops.sourceIsSame(sourceOffset, _offset); }
+				))
+					continue;
+				// Bring up the target slot that would otherwise become unreachable.
+				for (size_t targetOffset: ranges::views::iota(0u, _ops.targetSize()))
+					if (!_ops.targetIsArbitrary(targetOffset) && _ops.isCompatible(sourceOffset, targetOffset))
+					{
+						_ops.pushOrDupTarget(targetOffset);
+						return true;
+					}
+			}
+			// This slot needs to be moved.
+			else if (!_ops.isCompatible(sourceOffset, sourceOffset))
+			{
+				// If the current top fixes the slot, swap it down now.
+				if (_ops.isCompatible(_ops.sourceSize() - 1, sourceOffset))
+				{
+					_ops.swap(_ops.sourceSize() - sourceOffset - 1);
+					return true;
+				}
+				// Bring up a slot to fix this now, if possible.
+				if (bringUpTargetSlot(_ops, sourceOffset))
+					return true;
+				// Otherwise swap up the slot that swill fix the offending slot.
+				for (auto offset: ranges::views::iota(sourceOffset + 1, _ops.sourceSize()))
+					if (_ops.isCompatible(offset, sourceOffset))
+					{
+						_ops.swap(_ops.sourceSize() - offset - 1);
+						return true;
+					}
+				// Otherwise give up - we will need stack compression or stack limit evasion.
+			}
+		}
+		return false;
+	}
+	/// Finds a slot to dups or pushes with the aim of eventually fixing @a _targetOffset in the target.
+	/// In the simplest case, the slot at @a _targetOffset has a multiplicity > 0, i.e. it can directly be dupped or pushed
+	/// and the next iteration will fix @a _targetOffset.
+	/// But, in general, there may already be enough copies of the slot that is supposed to end up at @a _targetOffset
+	/// on stack, s.t. it cannot be dupped again. In that case there has to be a copy of the desired slot on stack already
+	/// elsewhere that is not yet in place (`nextOffset` below). The fact that ``nextOffset`` is not in place means that
+	/// we can (recursively) try bringing up the slot that is supposed to end up at ``nextOffset`` in the *target*.
+	/// When the target slot at ``nextOffset`` is fixed, the current source slot at ``nextOffset`` will be
+	/// at the stack top, which is the slot required at @a _targetOffset.
+	static bool bringUpTargetSlot(ShuffleOperations& _ops, size_t _targetOffset)
+	{
+		std::list<size_t> toVisit{_targetOffset};
+		std::set<size_t> visited;
+
+		while (!toVisit.empty())
+		{
+			auto offset = *toVisit.begin();
+			toVisit.erase(toVisit.begin());
+			visited.emplace(offset);
+			if (_ops.targetMultiplicity(offset) > 0)
+			{
+				_ops.pushOrDupTarget(offset);
+				return true;
+			}
+			// The desired target slot must already be somewhere else on stack right now.
```

Comment:
``bringUpTargetSlot`` and ``dupDeepSlotIfRequired`` are the weirdest and most complicated cases, I might actually move them down trying to make people start with the general algorithm first :-).
The comment here is only true because ``bringUpTargetSlot`` is only called in situations in which it's true.
Also "desired target slot" is probably not overly clear. The "desired target slot" is the best slot to dup/push right now.

First occurrence:
all source slots are compatible with target slots, but target is larger => there must be slots to dup/push.

Second occurrence:
We already moved the top away, whenever it's required elsewhere, or popped it, whenever it's no longer needed. So the current source is smaller or equally long as the target (actually, I should add an assertion about that, that might make it clearer). Furthermore a source slot should be removed. By the pigeonhole principle there has to be a target slot that needs with multiplicity>0.

Third occurrence:
The source is smaller than the target, i.e. there must be a target slot with multiplicity>0.

But yeah, maybe easier to talk through it in a call...